src/ucs/datastruct/ptr_map.inl

```diff
@@ -33,37 +33,233 @@ BEGIN_C_DECLS
+{
+    if (ucs_unlikely(ucs_ptr_map_key_indirect(key))) {
+        return 0;
+    }
+
+    *ptr_p = (void*)key;
+    return 1;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucs_ptr_map_init(ucs_ptr_map_t *map, int is_put_thread_safe,
+                 ucs_ptr_safe_hash_t *safe_hash)
 {
     UCS_STATIC_ASSERT(!ucs_ptr_map_key_indirect(UCS_PTR_MAP_KEY_INVALID));
     map->next_id = 0;
     kh_init_inplace(ucs_ptr_map_impl, &map->hash);
+
+    if (is_put_thread_safe) {
+        kh_init_inplace(ucs_ptr_map_impl, &safe_hash->hash);
+        ucs_spinlock_init(&safe_hash->lock, 0);
+    }
+
     return UCS_OK;
 }
 
-/**
- * Destroy a pointer map.
- *
- * @param [in]  map     Map to destroy.
- */
-static inline void ucs_ptr_map_destroy(ucs_ptr_map_t *map)
+static UCS_F_ALWAYS_INLINE void
+ucs_ptr_map_destroy(ucs_ptr_map_t *map, int is_put_thread_safe,
+                    ucs_ptr_safe_hash_t *safe_hash)
 {
     size_t size = kh_size(&map->hash);
 
+    if (is_put_thread_safe) {
+        size += kh_size(&safe_hash->hash);
+    }
+
     if (size != 0) {
         ucs_warn("ptr map %p contains %zd elements on destroy", map, size);
     }
-
     kh_destroy_inplace(ucs_ptr_map_impl, &map->hash);
+
+    if (is_put_thread_safe) {
+
+        kh_destroy_inplace(ucs_ptr_map_impl, &safe_hash->hash);
+        ucs_spinlock_destroy(&safe_hash->lock);
+    }
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucs_ptr_map_put(ucs_ptr_map_t *map, void *ptr, int indirect,
+                ucs_ptr_map_key_t *key, int is_put_thread_safe,
+                ucs_ptr_safe_hash_t *safe_hash)
+{
+    ucs_status_t status;
+    ucs_ptr_hash_t *hash;
+
+    if (ucs_likely(ucs_ptr_check_direct_put(ptr, !indirect, key))) {
+        return UCS_ERR_NO_PROGRESS;
+    }
+
+    if (is_put_thread_safe) {
+        hash = &safe_hash->hash;
+    } else {
+        hash = &map->hash;
+    }
+
+    if (is_put_thread_safe) {
+        ucs_spin_lock(&safe_hash->lock);
+    }
+
+    *key   = ucs_ptr_map_create_key(&map->next_id);
+    status = ucs_ptr_hash_put(hash, *key, ptr);
+
+    if (is_put_thread_safe) {
+        ucs_spin_unlock(&safe_hash->lock);
+    }
+
+    return status;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucs_ptr_map_get(ucs_ptr_map_t *map, ucs_ptr_map_key_t key, int extract,
+                void **ptr_p, int is_put_thread_safe,
+                ucs_ptr_safe_hash_t *safe_hash)
+{
+    ucs_status_t status;
+
+    if (ucs_likely(ucs_ptr_check_direct_get(key, ptr_p))) {
+        return UCS_ERR_NO_PROGRESS;
+    }
+
+    status = ucs_ptr_hash_get(&map->hash, key, extract, ptr_p);
+    if (ucs_likely(status != UCS_ERR_NO_ELEM)) {
+        return status;
+    }
+
+    if (is_put_thread_safe) {
+        ucs_spin_lock(&safe_hash->lock);
+        status = ucs_ptr_hash_get(&safe_hash->hash, key, 1, ptr_p);
+        ucs_spin_unlock(&safe_hash->lock);
+        if ((status == UCS_OK) && !extract) {
+            ucs_ptr_hash_put(&map->hash, key, *ptr_p);
```

Comment:
just to be sure, is this safe?

Reply:
Yes it is safe taking into account that we call this method simultaneously only with put.